edalize/nextpnr.py

```diff
@@ -0,0 +1,168 @@
+        (src_files, incdirs) = self._get_fileset_files()
+
+        yosys_synth_options = self.tool_options.get("yosys_synth_options", "")
+        yosys_template = self.tool_options.get("yosys_template", None)
+        yosys_edam = {
+                "files"         : self.files,
+                "name"          : self.name,
+                "toplevel"      : self.toplevel,
+                "parameters"    : self.parameters,
+                "tool_options"  : {
+                                    "yosys" : {
+                                        "arch" : "xilinx",
+                                        "yosys_synth_options" : yosys_synth_options,
+                                        "yosys_template" : yosys_template,
+                                        "yosys_as_subtool" : True,
+                                    }
+                                }
+                }
+
+        yosys = getattr(import_module("edalize.yosys"), "Yosys")(yosys_edam, self.work_root)
+        yosys.configure()
+
+        part_of_toolchain = self.tool_options.get("nextpnr_as_subtool", False)
+
+        file_table = []
+
+        output_format = self.tool_options.get("output_format", "fasm")
+        arch = self.tool_options.get("arch")
+
+        if not arch in getattr(self, "archs"):
+            logger.error('Missing or invalid "arch" parameter in "tool_options"')
+
+        synth_design = None
+        chipdb = None
+        device = None
+        xdc = None
+
+        for f in src_files:
+            if f.file_type in ["synthJson"]:
+                synth_design = f.name
+            elif f.file_type in ["bba"]:
+                chipdb = f.name
+            elif f.file_type in ["device"]:
+                device = f.name
+            elif f.file_type in ["xdc"]:
+                xdc = f.name
+            else:
+                continue
+
+        if chipdb is None:
+            logger.error("Missing required chipdb file")
+
+        if xdc is None:
+            logger.error("Missing required XDC file")
+
+        if device is None and arch is "fpga_interchange":
+            logger.error('Missing required ".device" file for "fpga_interchange" arch')
+
+        additional_options = self.tool_options.get("nextpnr_impl_options", "")
+
+        package = self.tool_options.get("package", None)
+        if package is None and arch is "fpga_interchange":
+            logger.error('Missing required "package" parameter for "fpga_interchange" arch')
+        package = package.split("-")[0] if arch == "fpga_interchange" else None
+
+        family = self.tool_options.get("family", None)
+        if family is None and arch is "fpga_interchange":
+            logger.error('Missing required "family" parameter for "fpga_interchange" arch')
+        if family not in getattr(self, "families") and arch is "fpga_interchange":
+            logger.error("Unsupported family: {}".format(family))
+
+        schema_dir = self.tool_options.get("schema_dir", None)
```

Comment:
If I understand correctly, schema_dir is system-specific rather than design-specific. I'm trying to the largest extent possible to not require users to supply system-specific paths in EDAM as that tends to lead to hard-coded paths. Is there some way we can query the system for the path to these schemas instead?

Reply:
I reworked this part to check if the `INTERCHANGE_SCHEMA_PATH` env variable is present -> if not it raises runtime error for `fpga_interchange` variant.